server/Arcadia.Assistant.ExternalStorages.SharepointOnline/SharepointStorage.cs

```diff
@@ -0,0 +1,249 @@
+                .ToArray();
+        }
+
+        public async Task<StorageItem> AddItem(string list, StorageItem item, CancellationToken cancellationToken)
+        {
+            var listItemType = await this.GetListItemType(list, cancellationToken);
+            var requestData = this.StorageItemToListItemRequest(item, listItemType);
+
+            var request = SharepointRequest
+                .Create(HttpMethod.Post, this.GetListItemsUrl(list))
+                .WithContent(requestData);
+
+            var addedItem = await this.ExecuteSharepointRequest<SharepointListItem>(request, cancellationToken);
+
+            return this.ListItemToStorageItem(addedItem);
+        }
+
+        public async Task UpdateItem(string list, StorageItem item, IEnumerable<ICondition> conditions, CancellationToken cancellationToken)
+        {
+            var listItems = (await this.GetListItems(list, conditions, cancellationToken)).ToArray();
+
+            this.EnsureSingleItemReturned(listItems);
+
+            var existingListItem = listItems.First();
+
+            var updateItemUrl = this.GetListItemsUrl(list, false);
+            updateItemUrl += $"({existingListItem.Id})";
+
+            var listItemType = await this.GetListItemType(list, cancellationToken);
+
+            var requestData = this.StorageItemToListItemRequest(item, listItemType);
+
+            var request = SharepointRequest
+                .Create(HttpMethod.Post, updateItemUrl)
+                .WithContent(requestData)
+                .WithIfMatchHeader()
+                .WithXHttpMethodHeader("MERGE");
+
+            await this.ExecuteSharepointRequest<SharepointListItem>(request, cancellationToken);
+        }
+
+        public async Task DeleteItem(string list, IEnumerable<ICondition> conditions = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var listItems = (await this.GetListItems(list, conditions, cancellationToken)).ToArray();
+
+            this.EnsureSingleItemReturned(listItems);
+
+            var existingListItem = listItems.First();
+
+            var deleteItemUrl = this.GetListItemsUrl(list, false);
+            deleteItemUrl += $"({existingListItem.Id})";
+
+            var request = SharepointRequest
+                .Create(HttpMethod.Post, deleteItemUrl)
+                .WithIfMatchHeader()
+                .WithXHttpMethodHeader("DELETE");
+
+            await this.ExecuteSharepointRequest(request, cancellationToken);
+        }
+
+        private async Task<IEnumerable<SharepointListItem>> GetListItems(
+            string list,
+            IEnumerable<ICondition> conditions,
+            CancellationToken cancellationToken)
+        {
+            var itemsUrl = this.GetListItemsUrl(list);
+
+            var conditionsUrlPart = this.conditionsCompiler.CompileConditions(conditions);
+
+            if (conditionsUrlPart != null)
+            {
+                itemsUrl += $"&{conditionsUrlPart}";
```

Comment:
uri builder?

Reply:
How do you suggest to use it? It doesn't has method that gets existing url and add query string parameter to it. Or I haven't found it.